include/dali/core/mm/binning_resource.h

```diff
@@ -0,0 +1,195 @@
+ * resource 2.
+ *
+ */
+template <typename Kind,
+          int nstatic_bins = -1,
+          typename Interface = mm::memory_resource<Kind>>
+class binning_resource_base : public Interface {
+ public:
+  using memory_kind = Kind;
+  static constexpr int static_num_bins = nstatic_bins;
+  static_assert(static_num_bins != 0);
+
+  /** @brief Constructs a binning resource from an array of split points, resources and extra data
+   *
+   * This constructor servers is used in a type deduction guide.
+   *
+   * @tparam SplitPointCollection   a collection of split points; it must have element type
+   *                                convertible to size_t and support iterationa and obtaining size
+   *                                with std::size
+   *
+   * @tparam ResourceCollection     a collection of pointers or iterators to memory_resource<Kind>
+   *                                objects; the elements must be convertible to a plain pointer
+   *                                by applying operators &* in succession
+   *
+   * @param split_points            a collection of split points; it must have one fewer element
+   *                                than resources
+   * @param resources               a collection of memory resource pointers or smart poitners or
+   *                                iterators
+   */
+  template <typename SplitPointCollection, typename ResourceCollection>
+  binning_resource_base(const SplitPointCollection &split_points,
+                        const ResourceCollection &resources) {
+    if constexpr (static_num_bins > 0) {
+      assert(split_points.size() == static_num_bins - 1);
+      assert(resources.size() == static_num_bins);
+    }
+
+    assert(dali::size(split_points) + 1 == dali::size(resources));
+    resize_if_possible(bin_split_points_, dali::size(split_points));
+    resize_if_possible(resources_, dali::size(resources));
+
+    int i = 0;
+    for (auto split_point : split_points) {
+      bin_split_points_[i] = split_point;
+      if (i && split_point <= bin_split_points_[i-1])
+        throw std::invalid_argument("Bin split points must be strictly increasing.");
+      i++;
+    }
+    assert(i == num_bins() - 1);
+
+    i = 0;
+    for (auto &&ptr : resources)
+      resources_[i++] = &*ptr;
+
+    assert(i == num_bins());
+  }
+
+  int num_bins() const {
+    return resources_.size();
+  }
+
+  size_t split_point(int after_bin) const {
+    assert(after_bin >= 0 && after_bin < num_bins() - 1);
+    return bin_split_points_[after_bin];
+  }
+
+  auto resource(int bin) const {
+    assert(bin >= 0 && bin < num_bins());
+    return resources_[bin];
+  }
+
+ private:
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    int bin = find_bin(bytes);
+    return resource(bin)->allocate(bytes, alignment);
+  }
+
+  void do_deallocate(void *ptr, size_t bytes, size_t alignment) override {
+    int bin = find_bin(bytes);
```

Comment:
We can rely on the fact that the deallocate will be called with the same number of bytes as the allocation? There are no allocation shrinking tricks possible here that would make us forward the call to wrong resource?

Reply:
No, partial deallocation is illegal. The size and alignment must match exactly - so that resources dealing with host-inaccessible memory don't need to maintain (as much) auxiliary information, including maps, etc.